secure/secure.sh

```diff
@@ -0,0 +1,403 @@
+  kadmin -p root -w $root_principal_password -q "xst -k mapred.keytab mapred/$FQDN HTTP/$FQDN"
+  kadmin -p root -w $root_principal_password -q "xst -k http.keytab HTTP/$FQDN"
+
+  # Move keytabs to /etc/hadoop/conf and set permissions on them
+  mv *.keytab $HADOOP_CONF_DIR
+  chown hdfs:hadoop $HADOOP_CONF_DIR/hdfs.keytab
+  chown yarn:hadoop $HADOOP_CONF_DIR/yarn.keytab
+  chown mapred:hadoop $HADOOP_CONF_DIR/mapred.keytab
+  chown hdfs:hadoop $HADOOP_CONF_DIR/http.keytab
+  chmod 400 $HADOOP_CONF_DIR/*.keytab
+}
+
+function set_property_in_xml() {
+  bdconfig set_property \
+    --configuration_file $1 \
+    --name $2 --value $3 \
+    --create_if_absent \
+    --clobber \
+    || err "Unable to set $2"
+}
+
+function set_property_core_site() {
+  set_property_in_xml "$HADOOP_CONF_DIR/core-site.xml" $1 $2
+}
+
+function config_core_site() {
+  set_property_core_site 'hadoop.security.authentication' 'kerberos'
+  set_property_core_site 'hadoop.security.authorization' 'true'
+  set_property_core_site 'hadoop.rpc.protection' 'privacy'
+  set_property_core_site 'hadoop.ssl.enabled' 'true'
+  set_property_core_site 'hadoop.ssl.require.client.cert' 'false'
+  set_property_core_site 'hadoop.ssl.hostname.verifier' 'DEFAULT'
+  set_property_core_site 'hadoop.ssl.keystores.factory.class' 'org.apache.hadoop.security.ssl.FileBasedKeyStoresFactory'
+  set_property_core_site 'hadoop.ssl.server.conf' 'ssl-server.xml'
+  set_property_core_site 'hadoop.ssl.client.conf' 'ssl-client.xml'
+  set_property_core_site 'fs.default.name' "hdfs://$MASTER_FQDN"
+  set_property_core_site 'fs.defaultFS' "hdfs://$MASTER_FQDN"
+}
+
+function set_property_hdfs_site() {
+  set_property_in_xml "$HADOOP_CONF_DIR/hdfs-site.xml" $1 $2
+}
+
+function config_hdfs_site() {
+  set_property_hdfs_site 'dfs.block.access.token.enable' 'true'
+  set_property_hdfs_site 'dfs.encrypt.data.transfer' 'true'
+  # Name Node
+  set_property_hdfs_site 'dfs.namenode.keytab.file' "$HADOOP_CONF_DIR/hdfs.keytab"
+  set_property_hdfs_site 'dfs.namenode.kerberos.principal' "hdfs/_HOST@$REALM"
+  set_property_hdfs_site 'dfs.namenode.kerberos.internal.spnego.principal' "HTTP/_HOST@$REALM"
+  # Secondary Name Node
+  set_property_hdfs_site 'dfs.secondary.namenode.keytab.file' "$HADOOP_CONF_DIR/hdfs.keytab"
+  set_property_hdfs_site 'dfs.secondary.namenode.kerberos.principal' "hdfs/_HOST@$REALM"
+  set_property_hdfs_site 'dfs.secondary.namenode.kerberos.internal.spnego.principal' "HTTP/_HOST@$REALM"
+  # Journal Node
+  set_property_hdfs_site 'dfs.journalnode.keytab.file' "$HADOOP_CONF_DIR/hdfs.keytab"
+  set_property_hdfs_site 'dfs.journalnode.kerberos.principal' "hdfs/_HOST@$REALM"
+  set_property_hdfs_site 'dfs.journalnode.kerberos.internal.spnego.principal' "HTTP/_HOST@$REALM"
+  # Data Node
+  # 'dfs.data.transfer.protection' Enable SASL, only supported after version 2.6
+  set_property_hdfs_site 'dfs.data.transfer.protection' 'privacy'
+  unset HADOOP_SECURE_DN_USER  # Required by SASL
```

Comment:
I don't think this is necessary unless you sourced `hadoop-env.sh`.

Reply:
Done.